### PR TITLE
[ONNX] Special post process for onnx::Cast and onnx::ConstantOfShape shape type inference

### PR DESCRIPTION
**Pull request by BowenBao**
* Special post process for onnx::Cast and onnx::ConstantOfShape 
* Update `test_pytorch_onnx_shape_inference.py` to be unit test over shape inference patterns.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_shape_inference.py [hunk 1]
```diff
@@ -1,72 +1,69 @@
...
 
-import test_pytorch_onnx_onnxruntime
-from test_pytorch_onnx_onnxruntime import TestONNXRuntime
-from torch.onnx import utils, OperatorExportTypes, TrainingMode
-from torch.onnx.utils import _validate_dynamic_axes
-from torch.onnx.symbolic_helper import (_set_opset_version, _set_operator_export_type,
-                                        _set_onnx_shape_inference, _set_training_mode,
-                                        _is_tensor_list, _is_tensor, _is_none)
+def expect_tensor(scalar_type, shape=None, dynamic_shape=None):
+    def verify(actual_type):
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_shape_inference.py:7]: Seems that there is only code in `expect_tensor` is `verify`.
Can we just use the code in `verify` instead of having this function?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 2]
```diff
@@ -1160,6 +1160,57 @@ void SpecialPostProcess(Node* n) {
...
+    }
+    case ::c10::onnx::ConstantOfShape: {
+      // ONNX shape inference is not able to propagate output tensor shape
+      // for onnx::ConstantOfShape if input `shape` is not constant.
+      // This is a temporary solution when some partial information is available,
+      // for example, knowing rank of output tensor, or knowing symbolic shape.
+      // This solution won't be needed once we have proper symbolic propagation.
+      auto shape_node = n->input(0)->node();
+      if (shape_node->kind() == ::c10::onnx::Shape) {
+        // shape -> ConstantOfShape
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1184]: nit: Shape

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 3]
```diff
@@ -1160,6 +1160,57 @@ void SpecialPostProcess(Node* n) {
...
+            n->output()->setType(v_type);
+          }
+        }
+      } else if (shape_node->kind() == ::c10::prim::ListConstruct) {
+        // ListConstruct -> ConstantOfShape
+        auto v_type = n->output()->type()->cast<TensorType>();
+        if (v_type && !v_type->sizes().concrete_sizes()) {
+          auto value = n->t(attr::value);
+          v_type = v_type->withScalarType(value.scalar_type());
+          std::vector<c10::ShapeSymbol> sizes;
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1204]: `std::vector<c10::ShapeSymbol> sizes(shape_node->inputs().size(), c10::ShapeSymbol::newSymbol());` ?

**Review by jiafatom — APPROVED** [review]
Some minor comments.

File: test/onnx/test_pytorch_onnx_shape_inference.py [hunk 4]
```diff
@@ -1,72 +1,69 @@
...
 
+    def test_constant_of_shape_dynamic(self):
+        # Test ConstantOfShape with input of prim::ListConstruct of dynamic tensor
+        rank = 4
+        g = self.create_empty_graph()
+        inputs = [g.addInput() for i in range(rank)]
+        shape = g.op("prim::ListConstruct", *inputs)
+        shape.setType(torch._C.ListType.ofInts())
+        constant_of_shape = g.op("ConstantOfShape", shape, value_t=torch.tensor([2.0]))
+        self.run_test(g, constant_of_shape.node(), expect_tensor('Float', dynamic_shape=(None, None, None, None)))
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_shape_inference.py:66]: Please add tests about reshape dim = 0 cases mentioned in https://github.com/pytorch/pytorch/pull/56053

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_shape_inference.py:66]: thx

File: test/onnx/test_pytorch_onnx_shape_inference.py (hunk 1 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_shape_inference.py:7]: `expect_tensor` returns a function object. `expect_tensor` serves as api for devs to pass in expected shape/dtype info. The returned function object takes actual type as argument, and applies test assertion.
